core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -75,14 +86,105 @@ public DnsResolverGroupBuilder resolvedAddressTypes(
+     * option when a <strong>limited</strong> number of hostnames are used.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder enableAutoRefresh(boolean autoRefresh) {
+        this.autoRefresh = autoRefresh;
+        return this;
+    }
+
+    /**
+     * Sets the {@link Backoff} which is used when the {@link DnsNameResolver} fails to update the cache.
+     */
+    public DnsResolverGroupBuilder autoRefreshBackoff(Backoff refreshBackoff) {
+        autoRefreshBackoff = requireNonNull(refreshBackoff, "refreshBackoff");
+        return this;
+    }
+
+    /**
+     * Sets the {@link ToLongFunction} which determines how long the {@link DnsRecord}s for a hostname should
+     * be refreshed after the first cache.
+     *
+     * <p>For example:
+     * <pre>{@code
+     * dnsResolverGroupBuilder.autoRefreshTimeout(hostname -> {
+     *     if (hostname.endsWith("busy.domain.com")) {
+     *         return Duration.ofDays(7).toMillis(); // Automatically refresh the cached domain for 7 days.
+     *     }
+     *     if (hostname.endsWith("sporadic.domain.dom")) {
+     *         return 0; // Don't need to refresh a sporadically used domain.
+     *     }
+     *     ...
+     * });
+     * }</pre>
+     *
+     * <p>Note this method is mutually exclusive with {@link #autoRefreshTimeout(Duration)} and
+     * {@link #autoRefreshTimeoutMillis(long)}.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(ToLongFunction<? super String> timeoutFunction) {
+        requireNonNull(timeoutFunction, "timeoutFunction");
+        checkState(autoRefreshTimeoutFunction == null, "'autoRefreshTimeout()' was set already: %s",
+                   autoRefreshTimeoutFunction);
+        //noinspection unchecked
+        autoRefreshTimeoutFunction = (ToLongFunction<String>) timeoutFunction;
+        return this;
+    }
+
+    /**
+     * Sets the timeout after which a refreshing {@link DnsRecord} should expire.
+     * If this option is unspecified and {@link #enableAutoRefresh(boolean)} is set to
+     * {@code true}, a cached {@link DnsRecord} is automatically refreshed until the {@link ClientFactory}
+     * is closed.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(Duration timeout) {
+        requireNonNull(timeout, "timeout");
+        return autoRefreshTimeoutMillis(timeout.toMillis());
+    }
+
+    /**
+     * Sets the timeout in milliseconds after which a refreshing {@link DnsRecord} should expire.
+     * If this option is unspecified and {@link #enableAutoRefresh(boolean)} is set to
+     * {@code true}, a cached {@link DnsRecord} is automatically refreshed until the {@link ClientFactory}
+     * is closed.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeoutMillis(long timeoutMillis) {
+        checkArgument(timeoutMillis > 0, "timeoutMillis: %s (expected: > 0)", timeoutMillis);
```

Comment:
How about also accepting `0` as a valid input as we do in other APIs?

Reply:
I intentionally excluded `0` because `0` is used to disable auto-refresh in the predication function which could conflict with `enableAutoRefresh(false)`.
https://github.com/line/armeria/blob/0db3824415f5255cb6022431d94d7604d34212e7/core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java#L123

Reply:
> because 0 is used to disable auto-refresh in the predication function

How about also changing the value to return `-1`? We use `0` to disable the timeout through our API so I think this violates our convetion.

Reply:
It was also my considering point. I hesitated to use `0` for disabling the timeout because it causes that `0` can have different meanings depending on the context.
But agree that the consistency, which 0 means to disable the timeout, is important. 

Hmm... now I'm leaning toward using 0 for disable timeout.
Let me update the Javadoc for the 0 input and 0 return.